File: julia/experiment.jl
module Experiment

include("agent.jl")
include("cartpoleenv.jl")

const episodesnum = 20000000
const stepsnum = Env.fps * 10

struct History
    states::Vector{Env.State}
    actions::Vector{Float64}
    rewards::Vector{Float64}
end

function newhistory()
    states = Vector{Env.State}(undef, stepsnum)
    actions = Vector{Float64}(undef, stepsnum)
    rewards = Vector{Float64}(undef, stepsnum)
    History(states, actions, rewards)
end

function run()
    returns = Vector{Float64}(undef, episodesnum)
    qtable = Agent.newqtable()
    agentparams = Agent.defaultparams

    for episode = 1:episodesnum
        hist = oneepisode(agentparams, qtable)
        returns[episode] = sum(hist.rewards)
    end

    returns, qtable
end

function test(qtable)
    hist = oneepisode(Agent.defaulttestparams, qtable)
    hist, qtable
end

function oneepisode(agentparams, qtable)
    hist = newhistory()

    s = Env.newstate()
    a = 0.0
    r = 0.0

    for step = 1:stepsnum
        hist.states[step] = s
        hist.actions[step] = a
        hist.rewards[step] = r

        a = Agent.action(agentparams, qtable, s)
        snext = Env.onestep(s, anext)
        r = Env.reward(snext, a)
        Agent.learn!(agentparams, qtable, s, a, r, snext)

        s = snext
    end

    hist
end

end

File: haskell/src/experiment.hs
module Experiment
    ( Experiments
    , run
    , test
    )
where

import qualified Env                           as E
import qualified Agent                         as A

episodesNum = 10000
stepsNum = E.fps * 10

data Experiments = Experiments {params :: A.Params
    , qtable :: A.Qtable
    , states :: [E.State]
    , actions :: [E.Action]
    , rewards :: [E.Reward]
    }


initExperiments :: IO Experiments
initExperiments = return Experiments { params  = A.defaultParams
                                     , qtable  = A.newQtable
                                     , states  = [E.initState]
                                     , actions = [0.0]
                                     , rewards = [0.0]
                                     }

resetExperiments :: Experiments -> IO Experiments
resetExperiments ex = return Experiments { params  = A.defaultParams
                                         , qtable  = qtable ex
                                         , states  = [E.initState]
                                         , actions = [0.0]
                                         , rewards = [0.0]
                                         }

makeTestExperiments :: Experiments -> IO Experiments
makeTestExperiments ex = return Experiments { params  = A.defaultTestParams
                                            , qtable  = qtable ex
                                            , states  = [E.initState]
                                            , actions = [0.0]
                                            , rewards = [0.0]
                                            }


run :: IO ([IO E.Reward], Experiments)
run = do
    let initEx  = initExperiments
    let loop    = iterate (>>= oneEpisode) initEx
    let returns = map (\ioEx -> sum . rewards <$> ioEx) (take episodesNum loop)
    ex <- loop !! (episodesNum - 1)
    return (returns, ex)



test :: Experiments -> IO Experiments
test ex = do
    testEx <- makeTestExperiments ex
    oneEpisode testEx


oneEpisode :: Experiments -> IO Experiments
oneEpisode ex = do
    let loop = iterate (>>= oneStep) $ resetExperiments ex
    loop !! (stepsNum - 1)


oneStep :: Experiments -> IO Experiments
oneStep ex = do
    let s       = head $ states ex
    let qtable' = qtable ex
    let params' = params ex
    a <- A.action params' qtable' s
    let sNext = E.oneStep s a
    let r     = E.reward sNext
    A.learn params' qtable' s a r sNext
    return Experiments { params  = params'
                       , qtable  = qtable'
                       , states  = sNext : states ex
                       , actions = a : actions ex
                       , rewards = r : rewards ex
                       }
